aea/helpers/profiling.py

```diff
@@ -86,56 +88,58 @@ class Profiling(Runnable):
-        objects_instances_to_count: List[Type] = None,
-        objects_created_to_count: List[Type] = None,
         output_function: Callable[[str], None] = lambda x: print(x, flush=True),
     ) -> None:
         """
         Init profiler.
 
         :param period: delay between profiling output in seconds.
-        :param objects_instances_to_count: object to count
-        :param objects_created_to_count: object created to count
+        :param types_to_track: object types to count
         :param output_function: function to display output, one str argument.
         """
         if period < 1:  # pragma: nocover
-            raise ValueError("Period should be at least 1 second!")
+            raise ValueError("Profiling frequency should be at least 1 second!")
         super().__init__(threaded=True)
         self._period = period
         self._start_ts = time.time()
-        self._objects_instances_to_count = objects_instances_to_count or []
-        self._objects_created_to_count = objects_created_to_count or []
+        self._types_to_track: List[Type] = types_to_track
         self._output_function = output_function
-        self._counter: Dict[Type, int] = Counter()
+        self.object_counts: Dict[Type, List[int]] = {
+            obj: [0, 0] for obj in types_to_track
+        }  # {object: [instances_created, instances_deleted]}
+        self.set_counters()
 
         if platform.system() != "Windows":
             tracemalloc.start()
 
     def set_counters(self) -> None:
-        """Modify obj.__new__ to count objects created created."""
+        """Modify __new__ and __del__ to count objects created created and destroyed."""
 
-        def make_fn(obj: Any) -> Callable:
+        def call_count(wrapped: Callable, index: int, obj: Any) -> Callable:
             orig_new = obj.__new__
-            # pylint: disable=protected-access  # type: ignore
 
-            @wraps(orig_new)
-            def new(*args: Any, **kwargs: Any) -> Callable:
-                self._counter[obj] += 1
+            @wraps(wrapped)
+            def wrapper(*args: Any, **kwargs: Any) -> Callable:
+                self.object_counts[obj][index] += 1
+                # Avoid TypeError: object.__new__() takes exactly one argument
                 if orig_new is object.__new__:
                     return orig_new(args[0])  # pragma: nocover
-                return orig_new(*args, **kwargs)  # pragma: nocover
+                return wrapped(*args, **kwargs)
 
-            return new
+            return wrapper
 
-        for obj in self._objects_created_to_count:
-            self._counter[obj] = 0
-            obj.__new__ = make_fn(obj)  # type: ignore
+        for t in self._types_to_track:
+            t.__new__ = call_count(t.__new__, 0, t)
+            # For some reason, if we don't init the destructor with an empty function, the next
+            # line will raise the exception: AttributeError: type object 'Message' has no attribute '__del__'
```

Comment:
just to clarify: If an object implements the __del__ method, Python calls the __del__ method right before the garbage collector destroys the object. Hence it is not required; it is a finalizer and not a destructor. https://docs.python.org/3/reference/datamodel.html?highlight=__del__#object.__del__